opencog/atomutils/FindUtils.cc

```diff
@@ -273,44 +273,39 @@ bool contains_atomtype(const HandleSeq& clauses, Type atom_type)
 	return false;
 }
 
-
-HandleSeq get_free_vars_in_tree(const Handle& tree)
+OrderedHandleSet get_free_variables(const Handle& h, int quotation_level)
 {
-	OrderedHandleSet varset;
-
-	std::function<void (const Handle&)> find_rec = [&](const Handle& h)
-	{
-		Type t = h->getType();
-		if (t == VARIABLE_NODE)
-		{
-			varset.insert(h);
-			return;
-		}
-
-		// XXX FIXME Add support for UNQUOTE_LINK
-		if (t == QUOTE_LINK) return;
-		if (classserver().isA(t, SCOPE_LINK))
-		    // // A sugary version of ImplicationLink can be a ScopeLink
-		    // // only if it has 3 arguments
-		    // and (not classserver().isA(t, IMPLICATION_LINK)
-		    //      or h->getArity() == 3))
-			return;
-
-		if (h->isLink())
-		{
-			for (const Handle& oh : h->getOutgoingSet())
-				find_rec(oh);
-		}
-	};
-
-	find_rec(tree);
+	Type t = h->getType();
 
-	return HandleSeq(varset.begin(), varset.end());
+	// Base cases
+	if (t == VARIABLE_NODE and quotation_level < 1)
+		return {h};
+	if (h->isNode())
+		return {};
+
+	// Recursive cases
+	OrderedHandleSet results;
+	OC_ASSERT(h->isLink());
+	if (t == QUOTE_LINK) quotation_level++;
+	else if (t == UNQUOTE_LINK) quotation_level--;
+	for (const Handle& ch : h->getOutgoingSet()) {
+		OrderedHandleSet ch_free_var = get_free_variables(ch, quotation_level);
+		results.insert(ch_free_var.begin(), ch_free_var.end());
+	}
+	// If the link was a scope link then remove the scoped
+	// variables from the free variables found.
```

Comment:
The algorithm processes that correctly, in other words it will say that your example above is not closed. That is because the removal only occurs over the variables inside the children of the scope link.

There was almost code doing that but not quite. Ffirst the `FindAtoms` class does ignore quoted variables, but not scoped variables. Second, there was code that checks whether a given variable is scoped or quoted, but no code actually fetching unscoped, unquoted variables.